src/rgw/driver/rados/topic_migration.cc

```diff
@@ -0,0 +1,247 @@
+  int r = driver->read_topics(tenant, topics, &topics_objv, y, dpp);
+  if (r == -ENOENT) {
+    return 0; // ok, someone else already migrated
+  }
+  if (r < 0) {
+    ldpp_dout(dpp, 1) << "failed to read v1 topics from " << topics_obj
+        << " with: " << cpp_strerror(r) << dendl;
+    return r;
+  }
+
+  constexpr bool exclusive = true; // don't overwrite any existing v2 metadata
+  for (const auto& [name, topic] : topics.topics) {
+    // write the v2 topic
+    RGWObjVersionTracker objv;
+    objv.generate_new_write_ver(dpp->get_cct());
+    r = driver->write_topic_v2(topic, exclusive, objv, y, dpp);
+    if (r == -EEXIST) {
+      continue; // ok, someone else already migrated
+    }
+    if (r < 0) {
+      ldpp_dout(dpp, 1) << "v1 topic migration for " << topic.name
+          << " failed: " << cpp_strerror(r) << dendl;
+      return r;
+    }
+  }
+
+  // remove the v1 topics metadata (this destroys the lock too)
+  r = driver->remove_topics(tenant, &topics_objv, y, dpp);
+  if (r == -ECANCELED) {
+    return 0; // ok, someone else already migrated
+  }
+  if (r < 0) {
+    ldpp_dout(dpp, 1) << "failed to remove migrated v1 topics obj "
+        << topics_obj << " with: " << cpp_strerror(r) << dendl;
+    return r;
+  }
+  return r;
+}
+
+} // anonymous namespace
+
+int migrate(const DoutPrefixProvider* dpp,
+            rgw::sal::RadosStore* driver,
+            boost::asio::io_context& context,
+            spawn::yield_context yield)
+{
+  auto y = optional_yield{context, yield};
+
+  ldpp_dout(dpp, 1) << "starting v1 topic migration.." << dendl;
+
+  librados::Rados* rados = driver->getRados()->get_rados_handle();
+  const rgw_pool& pool = driver->svc()->zone->get_zone_params().log_pool;
+  librados::IoCtx ioctx;
+  int r = rgw_init_ioctx(dpp, rados, pool, ioctx);
+  if (r < 0) {
+    ldpp_dout(dpp, 1) << "failed to initialize log pool for listing with: "
+        << cpp_strerror(r) << dendl;
+    return r;
+  }
+
+  // loop over all objects with oid prefix "pubsub."
+  auto filter = rgw::AccessListFilterPrefix(rgw::sal::pubsub_oid_prefix);
+  constexpr uint32_t max = 100;
+  std::string marker;
+  bool truncated = false;
+
+  std::vector<std::string> oids;
+  do {
+    oids.clear();
+    r = rgw_list_pool(dpp, ioctx, max, filter, marker, &oids, &truncated);
+    if (r == -ENOENT) {
+      r = 0;
+      break;
+    }
+    if (r < 0) {
+      ldpp_dout(dpp, 1) << "failed to list v1 topic metadata with: "
+          << cpp_strerror(r) << dendl;
+      return r;
+    }
+
+    for (const std::string& oid : oids) {
+      const auto obj = rgw_raw_obj{pool, oid};
+      if (oid.find(rgw::sal::pubsub_bucket_oid_infix) != oid.npos) {
+        ldpp_dout(dpp, 4) << "migrating v1 topics " << oid << dendl;
+        r = migrate_notification(dpp, y, driver, obj);
+        ldpp_dout(dpp, 4) << "migrating v1 topics " << oid << " completed with: "
+                          << ((r == 0)? "successful": cpp_strerror(r)) << dendl;
```

Comment:
because https://github.com/ceph/ceph/pull/55527 will reject requests to change topics or notifications during migration, i think we should make any fatal migration errors as obvious as possible. we should already be logging errors at log level 1 that include the oid and error code, but admins aren't always watching all of their radosgw logs

there's a global "cluster log" that we can write errors to that will be visible to commands like `ceph status`. there's a `rgw_clog_warn()` helper function in driver/rados/rgw_tools.cc that you can use for this. make sure these errors also contain both the oid and error code. call that on errors from `migrate_notification()` and `migrate_topics()`

Reply:
continue logging everything with lddp. just the fatal migration errors go to the cluster log

Reply:
done